File: Semester5/DSA/DSA Lab/Lab4/Stack.cpp
#include<iostream>
#include"Stack.h"
using namespace std;

Stack::Stack()
{
    capacity = 10;
    _top = -1;
    array = new int[capacity];
}
bool Stack:: isStackFull()
{
    if(_top == capacity -1)
    {
        return true;
    }
    return false;
}
bool Stack:: isStackEmpty()
{
    if(_top==-1){
        return true;
    }
    return false;
}
 void Stack::push(int val)
 {
    if(!isStackFull())
    {
        // if(top == -1)
        array[++_top] == val;
        // else
        // top +=1;
        // array[top] = val;
    }
    else
    throw "Stack is full";
 }

 int Stack::top()
 {
    // int temp{-1};
    if(isStackEmpty())
    {
        return _top;
    }
    else
    {
        return  array[_top];
    }

 }
 
 void Stack::pop()
 {
    if(!isStackEmpty())
    {
        //int temp = array[top];
        _top -= 1;
        //  temp;
    }
    else
    {
        throw "Stack is empty";
    }
 }

 Stack::~Stack()
 {
    delete [] array;
    array = nullptr;
 }

File: Semester5/WebTech_Labs/lab4/lab4/Stack.java
package lab4;

public class Stack {
    private int arr[];
    private int size;
    private int capacity;
    private int top;

    private void resize() {
        capacity *= 2;
        int newArr[] = new int[capacity];
        for (int i = 0; i < size; i++) {
            newArr[i] = arr[i];
        }
        arr = newArr;
    }

    public Stack() {
        capacity = 2;
        size = 0;
        top = -1;
        arr = new int[capacity];
    }

    public Stack(int iCapacity) {
        capacity = iCapacity;
        size = 0;
        top = -1;
        arr = new int[capacity];
    }

    public void push(int val) {
        if (size == capacity) {
            resize();
        }
        arr[++top] = val;
        size++;
    }

    public int pop() {
        if (size == 0) {
            System.out.println("Stack is empty Cannot pop.");
            return -1;
        }
        int value = arr[top--];
        size--;
        return value;
    }

    public static void main(String[] args) {
        Stack stack = new Stack();
        Stack stack1 = new Stack();

        stack.push(5);
        stack1.push(12);
        System.out.println("Pushed element to stack1: 12");
        stack.push(10);
        stack.push(15);

        System.out.println("Popped element: " + stack.pop());
        System.out.println("Popped element: " + stack.pop());
        System.out.println("Popped element: " + stack.pop());
        System.out.println("Popped element: " + stack.pop());
    }
}
